File: scripts/transcribe.mjs
#!/usr/bin/env node
/**
 * Transcribe MP3 files using OpenAI Whisper API
 * Node.js alternative to the Python transcribe.py script
 */

import OpenAI from "openai";
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ROOT_DIR = path.join(__dirname, "..");

// Load environment variables from .env.local
const envPath = path.join(ROOT_DIR, ".env.local");
if (fs.existsSync(envPath)) {
  const envContent = fs.readFileSync(envPath, "utf-8");
  envContent.split("\n").forEach((line) => {
    const match = line.match(/^([^#=]+)=(.*)$/);
    if (match) {
      process.env[match[1].trim()] = match[2].trim();
    }
  });
}

const openai = new OpenAI({ apiKey: process.env.OPENAI_API_KEY });

/**
 * Transcribe a single audio file
 */
async function transcribeFile(filePath, videoId) {
  console.log(`  Transcribing: ${path.basename(filePath)}`);

  const fileStream = fs.createReadStream(filePath);

  const response = await openai.audio.transcriptions.create({
    model: "whisper-1",
    file: fileStream,
    response_format: "verbose_json",
    timestamp_granularities: ["segment"],
  });

  return {
    source_file: filePath,
    video_id: videoId,
    title: `YouTube Video ${videoId}`,
    speaker: "Unknown",
    source_url: `https://www.youtube.com/watch?v=${videoId}`,
    topic_tags: ["entrepreneurship"],
    categories: ["Education"],
    transcript_text: response.text,
    segments: (response.segments || []).map((seg) => ({
      start: seg.start,
      end: seg.end,
      text: seg.text,
    })),
    duration_seconds: response.duration || 0,
  };
}

/**
 * Main function
 */
async function main() {
  const args = process.argv.slice(2);
  const limitArg = args.find((a) => a.startsWith("--limit="));
  const limit = limitArg ? parseInt(limitArg.split("=")[1]) : 5;

  const mp3Dir = path.join(ROOT_DIR, "mp3");
  const outputDir = path.join(ROOT_DIR, "content", "transcripts");

  // Ensure output directory exists
  if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir, { recursive: true });
  }

  // Get MP3 files
  if (!fs.existsSync(mp3Dir)) {
    console.error(`MP3 directory not found: ${mp3Dir}`);
    process.exit(1);
  }

  const mp3Files = fs.readdirSync(mp3Dir)
    .filter((f) => f.endsWith(".mp3"))
    .map((f) => path.join(mp3Dir, f));

  console.log(`Found ${mp3Files.length} MP3 files`);

  // Check which files already have transcripts
  const existingTranscripts = new Set(
    fs.readdirSync(outputDir)
      .filter((f) => f.endsWith(".json"))
      .map((f) => f.replace(".json", ""))
  );

  // Filter to files without transcripts AND under 25MB (Whisper limit)
  const MAX_SIZE = 25 * 1024 * 1024; // 25MB
  const filesToProcess = mp3Files
    .filter((f) => {
      const name = path.basename(f, ".mp3");
      if (existingTranscripts.has(name)) return false;
      const size = fs.statSync(f).size;
      if (size > MAX_SIZE) {
        console.log(`Skipping ${name} (${(size / 1024 / 1024).toFixed(1)}MB > 25MB limit)`);
        return false;
      }
      return true;
    })
    .slice(0, limit);

  console.log(`Already transcribed: ${existingTranscripts.size}`);
  console.log(`Will transcribe: ${filesToProcess.length} files\n`);

  if (filesToProcess.length === 0) {
    console.log("No new files to transcribe!");
    return;
  }

  // Estimate cost
  let totalSize = 0;
  for (const f of filesToProcess) {
    totalSize += fs.statSync(f).size;
  }
  const estimatedMinutes = totalSize / (1024 * 1024) * 2; // Rough estimate
  const estimatedCost = estimatedMinutes * 0.006;
  console.log(`Estimated cost: ~$${estimatedCost.toFixed(2)}\n`);

  // Process files
  let successful = 0;
  const failed = [];

  for (let i = 0; i < filesToProcess.length; i++) {
    const filePath = filesToProcess[i];
    const videoId = path.basename(filePath, ".mp3");

    console.log(`[${i + 1}/${filesToProcess.length}] Processing ${videoId}...`);

    try {
      const result = await transcribeFile(filePath, videoId);

      // Save transcript
      const outputPath = path.join(outputDir, `${videoId}.json`);
      fs.writeFileSync(outputPath, JSON.stringify(result, null, 2));

      console.log(`  ✓ Saved (${Math.round(result.duration_seconds / 60)} min)\n`);
      successful++;
    } catch (error) {
      console.error(`  ✗ Failed: ${error.message}\n`);
      failed.push(videoId);
    }
  }

  console.log(`\n========================================`);
  console.log(`Completed: ${successful}/${filesToProcess.length}`);
  if (failed.length > 0) {
    console.log(`Failed: ${failed.join(", ")}`);
  }
  console.log(`Total transcripts: ${existingTranscripts.size + successful}`);
}

main().catch((err) => {
  console.error("Error:", err);
  process.exit(1);
});

File: scripts/transcribe.py
#!/usr/bin/env python3
"""
Transcribe MP3 files using OpenAI Whisper API.
Reads metadata from content_metadata.json and outputs transcript JSONs.
"""

import os
import json
from pathlib import Path
from openai import OpenAI
from dotenv import load_dotenv
from tqdm import tqdm

# Load environment variables
load_dotenv()


def transcribe_audio(client: OpenAI, audio_path: str, metadata: dict) -> dict:
    """
    Transcribe a single audio file using Whisper API.

    Args:
        client: OpenAI client
        audio_path: Path to MP3 file
        metadata: Dict containing video metadata

    Returns:
        Dict with transcript and metadata
    """
    with open(audio_path, "rb") as audio_file:
        transcript = client.audio.transcriptions.create(
            model="whisper-1",
            file=audio_file,
            response_format="verbose_json",
            timestamp_granularities=["segment"]
        )

    return {
        "source_file": str(audio_path),
        "video_id": metadata.get("video_id", ""),
        "title": metadata.get("title", "Unknown"),
        "speaker": metadata.get("channel", "Unknown"),
        "source_url": metadata.get("url", ""),
        "topic_tags": metadata.get("tags", [])[:10],  # Limit tags
        "categories": metadata.get("categories", []),
        "transcript_text": transcript.text,
        "segments": [
            {
                "start": seg.get("start", 0),
                "end": seg.get("end", 0),
                "text": seg.get("text", "")
            }
            for seg in (transcript.segments or [])
        ],
        "duration_seconds": transcript.duration
    }


def process_batch(
    mp3_dir: str,
    metadata_file: str,
    output_dir: str,
    limit: int = None,
    skip_existing: bool = True
):
    """
    Process a batch of audio files.

    Args:
        mp3_dir: Directory containing MP3 files
        metadata_file: Path to content_metadata.json
        output_dir: Where to save transcript JSONs
        limit: Maximum number of files to process (None = all)
        skip_existing: Skip files that already have transcripts
    """
    client = OpenAI()  # Uses OPENAI_API_KEY env variable

    # Load metadata
    with open(metadata_file, "r", encoding="utf-8") as f:
        metadata_data = json.load(f)

    videos = metadata_data.get("videos", [])
    print(f"Found {len(videos)} videos in metadata")

    # Create output directory
    output_path = Path(output_dir)
    output_path.mkdir(parents=True, exist_ok=True)

    # Build lookup by filename
    metadata_by_filename = {v["filename"]: v for v in videos}

    # Get MP3 files to process
    mp3_path = Path(mp3_dir)
    mp3_files = list(mp3_path.glob("*.mp3"))

    # Filter to files with metadata
    mp3_files = [f for f in mp3_files if f.name in metadata_by_filename]
    print(f"Found {len(mp3_files)} MP3 files with metadata")

    # Apply limit
    if limit:
        mp3_files = mp3_files[:limit]
        print(f"Processing first {limit} files")

    # Skip existing if requested
    if skip_existing:
        existing = set(f.stem for f in output_path.glob("*.json"))
        mp3_files = [f for f in mp3_files if f.stem not in existing]
        print(f"Skipping {len(existing)} already transcribed files")

    print(f"Will transcribe {len(mp3_files)} files")

    if not mp3_files:
        print("No files to process!")
        return

    # Estimate cost
    total_duration = sum(
        metadata_by_filename.get(f.name, {}).get("duration_seconds", 0)
        for f in mp3_files
    )
    estimated_cost = (total_duration / 60) * 0.006
    print(f"Estimated duration: {total_duration/60:.1f} minutes")
    print(f"Estimated cost: ${estimated_cost:.2f}")

    # Confirm
    response = input("\nProceed with transcription? (y/n): ")
    if response.lower() != "y":
        print("Aborted.")
        return

    # Process files
    successful = 0
    failed = []

    for mp3_file in tqdm(mp3_files, desc="Transcribing"):
        metadata = metadata_by_filename[mp3_file.name]

        try:
            result = transcribe_audio(client, str(mp3_file), metadata)

            output_file = output_path / f"{mp3_file.stem}.json"
            with open(output_file, "w", encoding="utf-8") as f:
                json.dump(result, f, indent=2, ensure_ascii=False)

            successful += 1

        except Exception as e:
            print(f"\nError transcribing {mp3_file.name}: {e}")
            failed.append(mp3_file.name)

    print(f"\n\nCompleted!")
    print(f"Successful: {successful}/{len(mp3_files)}")
    if failed:
        print(f"Failed: {len(failed)}")
        print("Failed files:", failed)


def main():
    import argparse

    parser = argparse.ArgumentParser(
        description="Transcribe MP3 files using OpenAI Whisper"
    )
    parser.add_argument(
        "--mp3-dir",
        default="./mp3",
        help="Directory containing MP3 files (default: ./mp3)"
    )
    parser.add_argument(
        "--metadata",
        default="./content_metadata.json",
        help="Metadata JSON file (default: ./content_metadata.json)"
    )
    parser.add_argument(
        "--output",
        default="./content/transcripts",
        help="Output directory for transcripts (default: ./content/transcripts)"
    )
    parser.add_argument(
        "--limit",
        type=int,
        default=None,
        help="Limit number of files to process"
    )
    parser.add_argument(
        "--no-skip",
        action="store_true",
        help="Don't skip already transcribed files"
    )

    args = parser.parse_args()

    process_batch(
        mp3_dir=args.mp3_dir,
        metadata_file=args.metadata,
        output_dir=args.output,
        limit=args.limit,
        skip_existing=not args.no_skip
    )


if __name__ == "__main__":
    main()
